src/node.cc

```diff
@@ -1136,14 +1140,77 @@ void PromiseRejectCallback(PromiseRejectMessage message) {
   callback->Call(process, arraysize(args), args);
 }
 
+Local<Value> GetPromiseReason(Environment* env, Local<Value> promise) {
+  Local<Function> fn = env->promise_unhandled_rejection();
+
+  Local<Value> internalProps =
+      Debug::GetInternalProperties(env->isolate(),
+                                   promise).ToLocalChecked().As<Value>();
+
+  // If fn is empty we'll almost certainly have to panic anyways
+  return fn->Call(env->context(), Null(env->isolate()), 1,
+                       &internalProps).ToLocalChecked();
+}
+
+void TrackPromise(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsObject());
+  Local<Object> promise = args[0].As<Object>();
+
+  TrackPromise::New(env->isolate(), promise);
+
+  Local<Value> err = GetPromiseReason(env, promise);
+  Local<NativeWeakMap> unhandled_reject_map =
+      env->promise_unhandled_reject_map();
+  Local<Set> unhandled_reject_keys =
+      env->promise_unhandled_reject_keys();
+
+  if (unhandled_reject_keys->AsArray()->Length() > 1000) {
+    return;
+  }
+
+  if (!unhandled_reject_map->Has(err) && !err->IsUndefined()) {
```

Comment:
Also, is there a point in checking whether the map already contains the value?  You can just set it and be done with it, right?